### PR TITLE
Add exclusions function to Windows Defender plugin

### PR DESCRIPTION
**Pull request by MaxGroot**
This PR adds functionality to retrieve exclusions added to Windows Defender  from the 'Exclusions' registry key.

### CODE REVIEW

File: dissect/target/plugins/os/windows/defender.py [hunk 1]
```diff
@@ -72,11 +72,18 @@
...
+DEFENDER_EXCLUSION_KEY = "HKLM\\SOFTWARE\\Microsoft\\Windows Defender\\Exclusions"
+
 DefenderLogRecord = TargetRecordDescriptor(
     "filesystem/windows/defender/evtx",
     [("datetime", "ts")] + DEFENDER_EVTX_FIELDS,
 )
 
+DefenderExclusionRecord = TargetRecordDescriptor(
+    "filesystem/windows/defender/exclusion",
+    [("datetime", "regf_mtime"), ("string", "type"), ("string", "value")],
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:84]:
```suggestion
    [
        ("datetime", "regf_mtime"),
        ("string", "type"),
        ("string", "value"),
    ],
```

File: dissect/target/plugins/os/windows/defender.py [hunk 2]
```diff
@@ -441,6 +450,27 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
                 else:
                     self.target.log.warning("Unknown Defender Detection Type %s", resource.detection_type)
 
+    @plugin.export(record=DefenderExclusionRecord)
+    def exclusions(self) -> Generator[Record, None, None]:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:454]:
```suggestion
    def exclusions(self) -> Iterator[DefenderExclusionRecord]:
```

File: dissect/target/plugins/os/windows/defender.py [hunk 3]
```diff
@@ -441,6 +450,27 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
...
+                # what is being excluded from Defender (e.g. powershell.exe, notepad.txt)
+                for exclusion in exclusion_type_subkey.values():
+                    exclusion_value = exclusion.name
+                    exclusion_type = exclusion_type_subkey.name
+                    # Due to the fact that every exclusion is a registry value and not a registry key, we can only know
+                    # the last modified timestamp of the exclusion type for a given exclusion, not a timestamp for the
+                    # exclusion itself. We reflect this to the analyst by using the regf_mtime field.
+                    exclusion_type_last_modified = exclusion_type_subkey.timestamp
+                    yield DefenderExclusionRecord(
+                        regf_mtime=exclusion_type_last_modified, type=exclusion_type, value=exclusion_value
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:471]:
```suggestion
                        regf_mtime=exclusion_type_subkey.timestamp,
                        type=exclusion_type,
                        value=exclusion_value,
                        _target=self.target,
```

File: dissect/target/plugins/os/windows/defender.py [hunk 4]
```diff
@@ -441,6 +450,27 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
...
+            for exclusion_type_subkey in exclusions_registry_key.subkeys():
+                # Every value is an exclusion for a said type. The 'name' property of said registry value holds
+                # what is being excluded from Defender (e.g. powershell.exe, notepad.txt)
+                for exclusion in exclusion_type_subkey.values():
+                    exclusion_value = exclusion.name
+                    exclusion_type = exclusion_type_subkey.name
+                    # Due to the fact that every exclusion is a registry value and not a registry key, we can only know
+                    # the last modified timestamp of the exclusion type for a given exclusion, not a timestamp for the
+                    # exclusion itself. We reflect this to the analyst by using the regf_mtime field.
+                    exclusion_type_last_modified = exclusion_type_subkey.timestamp
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:469]:
```suggestion
```

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/defender.py [hunk 5]
```diff
@@ -451,7 +455,7 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
                     self.target.log.warning("Unknown Defender Detection Type %s", resource.detection_type)
 
     @plugin.export(record=DefenderExclusionRecord)
-    def exclusions(self) -> Generator[Record, None, None]:
+    def exclusions(self) -> Generator[Record]:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:458]:
```suggestion
    def exclusions(self) -> Iterator[Record]:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/defender.py [hunk 6]
```diff
@@ -455,7 +455,7 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
                     self.target.log.warning("Unknown Defender Detection Type %s", resource.detection_type)
 
     @plugin.export(record=DefenderExclusionRecord)
-    def exclusions(self) -> Generator[Record]:
+    def exclusions(self) -> Iterator[Record]:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py]:
```suggestion
    def exclusions(self) -> Iterator[DefenderExclusionRecord]:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

File: tests/test_plugins_os_windows_defender.py [hunk 7]
```diff
@@ -108,3 +109,55 @@ def test_defender_quarantine_recovery(target_win, fs_win, tmp_path):
...
+    exclusions_key = VirtualKey(hive_hklm, "Software\\Microsoft\\Windows Defender\\Exclusions")
+    for exclusion_type, exclusions in exclusions_example.items():
+        exclusion_type_key = VirtualKey(hive_hklm, exclusions_key.path + f"\\{exclusion_type}")
+        for exclusion in exclusions:
+            exclusion_type_key.add_value(exclusion, 0)
+        exclusions_key.add_subkey(exclusion_type, exclusion_type_key)
+
+    # Prepare the mock target
+    hive_hklm.map_key(exclusions_key.path, exclusions_key)
+    target_win.registry.map_hive("HKEY_LOCAL_MACHINE", hive_hklm)
```
Comment (Schamper) [inline tests/test_plugins_os_windows_defender.py:136]:
```suggestion
def test_defender_exclusions(target_win, hive_hklm):
    # https://learn.microsoft.com/en-us/exchange/antispam-and-antimalware/windows-antivirus-software?view=exchserver-2019
    exclusions_example = {
        "Extensions": [".config", ".log", ".cfg"],
        "Processes": ["UmService.exe", "UmWorkerProcess.exe"],
        "Paths": ["C:\\System32\\Cluster\\"],
        "IpAddresses": [],
        "TemporaryPaths": [],
    }

    # Recreate the 'Exclusions' registry key based on the example dict
    exclusions_key = VirtualKey(hive_hklm, "Software\\Microsoft\\Windows Defender\\Exclusions")
    for exclusion_type, exclusions in exclusions_example.items():
        exclusion_type_key = VirtualKey(hive_hklm, exclusions_key.path + f"\\{exclusion_type}")
        for exclusion in exclusions:
            exclusion_type_key.add_value(exclusion, 0)
        exclusions_key.add_subkey(exclusion_type, exclusion_type_key)

```
If you add the `hive_hklm` fixture, you get the one that's already mapped into the `target_win` fixture.

Comment (Schamper) [reply tests/test_plugins_os_windows_defender.py:136]: Also, we started adding type hints to the test functions, could you add those?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/defender.py [hunk 8]
```diff
@@ -441,6 +454,28 @@ def quarantine(self) -> Iterator[Union[DefenderFileQuarantineRecord, DefenderBeh
                 else:
                     self.target.log.warning("Unknown Defender Detection Type %s", resource.detection_type)
 
+    @plugin.export(record=DefenderExclusionRecord)
+    def exclusions(self) -> Iterator[DefenderExclusionRecord]:
+        """Yield Microsoft Defender exclusions from the Registry"""
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/defender.py:459]:
```suggestion
        """Yield Microsoft Defender exclusions from the Registry."""
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — DISMISSED** [review]

File: dissect/target/plugins/os/windows/defender.py [hunk 9]
```diff
@@ -373,12 +384,14 @@ class MicrosoftDefenderPlugin(plugin.Plugin):
...
     def check_compatible(self):
-        # Either the Defender log folder or the quarantine folder has to exist for this plugin to be compatible.
+        # Either the Defender log folder, the quarantine folder or the registry
+        # has to exist for this plugin to be compatible.
 
         return any(
             [
                 self.target.fs.path(DEFENDER_LOG_DIR).exists(),
                 self.target.fs.path(DEFENDER_QUARANTINE_DIR).exists(),
+                self.target.has_function("registry"),
```
Comment (pyrco) [inline dissect/target/plugins/os/windows/defender.py:394]: Not just the registry, but the `DEFENDER_EXCLUSION_KEY` should also exist

**Review by pyrco — APPROVED** [review]

**Pull request merged**